contrib/db_pools/lib/src/pool.rs

```diff
@@ -0,0 +1,287 @@
+use rocket::async_trait;
+use rocket::{Build, Rocket};
+
+use crate::{Config, Error};
+
+/// This trait is implemented on connection pool types that can be used with the
+/// [`Database`] derive macro.
+///
+/// `Pool` determines how the connection pool is initialized from configuration,
+/// such as a connection string and optional pool size, along with the returned
+/// `Connection` type.
+///
+/// Implementations of this trait should use `async_trait`.
+///
+/// ## Example
+///
+/// ```
+/// use rocket::{Build, Rocket};
+///
+/// #[derive(Debug)]
+/// struct Error { /* ... */ }
+/// # impl std::fmt::Display for Error {
+/// #     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// #         unimplemented!("example")
+/// #     }
+/// # }
+/// # impl std::error::Error for Error { }
+///
+/// struct Pool { /* ... */ }
+/// struct Connection { /* .. */ }
+///
+/// #[rocket::async_trait]
+/// impl rocket_db_pools::Pool for Pool {
+///     type Connection = Connection;
+///     type InitError = Error;
+///     type GetError = Error;
+///
+///     async fn initialize(db_name: &str, rocket: &Rocket<Build>)
+///         -> Result<Self, rocket_db_pools::Error<Self::InitError>>
+///     {
+///         unimplemented!("example")
+///     }
+///
+///     async fn get(&self) -> Result<Connection, Self::GetError> {
+///         unimplemented!("example")
+///     }
+/// }
+/// ```
+#[async_trait]
+pub trait Pool: Sized + Send + Sync + 'static {
+    /// The type returned by get().
+    type Connection;
+
+    /// The error type returned by `initialize`.
+    type InitError: std::error::Error;
+
+    /// The error type returned by `get`.
+    type GetError: std::error::Error;
+
+    /// Constructs a pool from a [Value](rocket::figment::value::Value).
+    ///
+    /// It is up to each implementor of `Pool` to define its accepted
+    /// configuration value(s) via the `Config` associated type.  Most
+    /// integrations provided in `rocket_db_pools` use [`Config`], which
+    /// accepts a (required) `url` and an (optional) `pool_size`.
+    ///
+    /// ## Errors
+    ///
+    /// This method returns an error if the configuration is not compatible, or
+    /// if creating a pool failed due to an unavailable database server,
+    /// insufficient resources, or another database-specific error.
+    async fn initialize(db_name: &str, rocket: &Rocket<Build>)
+        -> Result<Self, Error<Self::InitError>>;
```

Comment:
This shouldn't be formatted this way. At worst:
```suggestion
    async fn initialize(
        db_name: &str,
        rocket: &Rocket<Build>
    ) -> Result<Self, Error<Self::InitError>>;
```

But I really think this needs to be `Self::Error`.